issue_4/contracts/GreenVault.sol

```diff
@@ -11,50 +22,82 @@ import "@yield-protocol/utils-v2/contracts/token/ERC20.sol";
 @notice A (single-token) vault that mints wrapped tokens
 in response to deposits
  */
-contract GreenVault is ERC20("GreenVaultToken", "OWL_GVT", 18) {
+contract GreenVault is
+    ERC20("GreenVaultToken", "OWL_GVT", 18),
+    Ownable,
+    FPMath
+{
     /**
     @notice the only token this vault supports
      */
-    IERC20 public theOnlySupportedToken;
+    IERC20Metadata public theOnlySupportedToken;
+
+    bool private is_entered;
+
+    /**
+    Multiplier: for each token deposited we mint `rewardFractionWad` Vault Tokens
+
+    It's represented as a wad (a 18-digit fixed point number)
+     */
+    uint256 public rewardFractionWad;
 
     /// @notice emitted on deposit
     event EventMint(address user, address token, uint256 amount);
     /// @notice emitted on withdrawal
     event EventBurn(address user, address token, uint256 amount);
 
+    event RewardFractionSet(uint256 rewardFractionWad);
+
+    modifier nonReentrant() {
+        require(!is_entered);
+        is_entered = true;
+        _;
+        is_entered = false;
+    }
+
     /**
     @param token address of the only token this vault will support
+    @param _rewardFractionWad reward fraction to start with
      */
-    constructor(IERC20 token) {
+    constructor(IERC20Metadata token, uint256 _rewardFractionWad)
+    {
         theOnlySupportedToken = token;
+        setRewardFraction(_rewardFractionWad);
+    }
+
+    function setRewardFraction(uint256 _rewardFractionWad) public onlyOwner {
+        rewardFractionWad = _rewardFractionWad;
+        emit RewardFractionSet(rewardFractionWad);
     }
 
     /**
     @notice Mint Vault token by depositing external token
     @param amount amount of tokens to deposit
      */
-    function mint(uint256 amount) public {
+    function deposit(uint256 amount) public nonReentrant {
         IERC20 token = theOnlySupportedToken;
 
-        _mint(msg.sender, amount);
+        uint256 rewardAmount = wadMul(amount, rewardFractionWad);
 
+        _mint(msg.sender, rewardAmount);
         token.transferFrom(msg.sender, address(this), amount);
 
-        emit EventMint(msg.sender, address(token), amount);
+        emit EventMint(msg.sender, address(token), rewardAmount);
```

Comment:
If you are always going to have `msg.sender` in the event as the user, then you don't need to make it explicit. `msg.sender` is always in the transaction logs.

Reply:
neat!

Reply:
(I have to point out that this won't scale beyond this toy contract, once the Vault itself or somebody else start minting tokens for other users)